File: packages/mobile-ios/Sources/RTCExpress/RTCExpress.swift
import Foundation
import WebRTC

public protocol RTCExpressDelegate: AnyObject {
    func rtcExpress(_ rtc: RTCExpress, didConnect userId: String)
    func rtcExpressDidDisconnect(_ rtc: RTCExpress)
    func rtcExpress(_ rtc: RTCExpress, didJoinRoom roomId: String, members: [String])
    func rtcExpress(_ rtc: RTCExpress, didReceive message: RoomMessage)
    func rtcExpress(_ rtc: RTCExpress, didReceive invite: CallInvite)
    func rtcExpress(_ rtc: RTCExpress, callState: CallStateUpdate)
    func rtcExpress(_ rtc: RTCExpress, remoteVideoTrack: RTCVideoTrack)
    func rtcExpress(_ rtc: RTCExpress, didJoinVoiceRoom roomId: String)
    func rtcExpress(_ rtc: RTCExpress, didLeaveVoiceRoom roomId: String)
    func rtcExpress(_ rtc: RTCExpress, didJoinVideoRoom roomId: String)
    func rtcExpress(_ rtc: RTCExpress, didLeaveVideoRoom roomId: String)
    func rtcExpressDidStartRecording(_ rtc: RTCExpress)
    func rtcExpress(_ rtc: RTCExpress, didStopRecording result: RecordingResult)
    func rtcExpress(_ rtc: RTCExpress, didError message: String)
}

public extension RTCExpressDelegate {
    func rtcExpress(_ rtc: RTCExpress, didConnect userId: String) {}
    func rtcExpressDidDisconnect(_ rtc: RTCExpress) {}
    func rtcExpress(_ rtc: RTCExpress, didJoinRoom roomId: String, members: [String]) {}
    func rtcExpress(_ rtc: RTCExpress, didReceive message: RoomMessage) {}
    func rtcExpress(_ rtc: RTCExpress, didReceive invite: CallInvite) {}
    func rtcExpress(_ rtc: RTCExpress, callState: CallStateUpdate) {}
    func rtcExpress(_ rtc: RTCExpress, remoteVideoTrack: RTCVideoTrack) {}
    func rtcExpress(_ rtc: RTCExpress, didJoinVoiceRoom roomId: String) {}
    func rtcExpress(_ rtc: RTCExpress, didLeaveVoiceRoom roomId: String) {}
    func rtcExpress(_ rtc: RTCExpress, didJoinVideoRoom roomId: String) {}
    func rtcExpress(_ rtc: RTCExpress, didLeaveVideoRoom roomId: String) {}
    func rtcExpressDidStartRecording(_ rtc: RTCExpress) {}
    func rtcExpress(_ rtc: RTCExpress, didStopRecording result: RecordingResult) {}
    func rtcExpress(_ rtc: RTCExpress, didError message: String) {}
}

public final class RTCExpress: SignalingListener {
    public weak var delegate: RTCExpressDelegate?
    public weak var incomingCallNotifier: IncomingCallNotifier?

    private let signaling = SignalingClient()
    private let callRecorder = CallRecorder()
    private var p2p: P2pMediaEngine?
    private var sfu: SfuMediaEngine?

    private var serverUrl = ""
    private var appId = ""
    private var token = ""
    private var userId = ""
    private var roomId: String?
    private var mediaModePref = "p2p"
    private var sfuUrl: String?
    private var inVoiceRoom = false
    private var inVideoRoom = false

    private struct ActiveCall {
        let callId: String
        let peerUserId: String
        let roomId: String
        let isCaller: Bool
        let callType: String
    }

    private var activeCall: ActiveCall?

    public init() {
        signaling.listener = self
    }

    public func initClient(_ options: RTCInitOptions) {
        serverUrl = options.serverUrl.trimmingCharacters(in: .init(charactersIn: "/"))
        appId = options.appId
        userId = options.userId
        token = options.token
        mediaModePref = options.mediaMode
        signaling.connect(serverUrl: serverUrl, token: options.token)
        Task {
            if let config = try? await PlatformApi.fetchPlatformConfig(serverUrl: serverUrl, appId: appId) {
                sfuUrl = config.sfuUrl
            }
        }
    }

    public func joinRoom(_ roomId: String) {
        self.roomId = roomId
        try? signaling.send(type: "join_room", payload: ["roomId": roomId])
    }

    public func sendMessage(_ text: String) throws {
        guard let roomId else { throw NSError(domain: "RTCExpress", code: 3) }
        try signaling.send(type: "send_message", payload: ["roomId": roomId, "text": text])
    }

    public func getMessageHistory(_ roomId: String, before: String? = nil, limit: Int = 50) async throws -> MessageHistoryPage {
        try await PlatformApi.getMessageHistory(serverUrl: serverUrl, token: token, roomId: roomId, before: before, limit: limit)
    }

    public func joinVoiceRoom() async throws {
        guard let roomId else { throw NSError(domain: "RTCExpress", code: 3) }
        guard let sfuUrl else { throw NSError(domain: "RTCExpress", code: 32, userInfo: [NSLocalizedDescriptionKey: "SFU not available"]) }
        if sfu == nil {
            sfu = SfuMediaEngine(sfuUrl: sfuUrl, userId: userId, authToken: token) { [weak self] type, payload in
                try self?.signaling.send(type: type, payload: payload)
            }
        }
        try await sfu?.joinRoom(roomId: roomId, options: SfuJoinOptions(audio: true, video: false, announceToRoom: true))
        inVoiceRoom = true
        try signaling.send(type: "join_media", payload: ["roomId": roomId, "kind": "voice"])
        delegate?.rtcExpress(self, didJoinVoiceRoom: roomId)
    }

    public func leaveVoiceRoom() throws {
        guard inVoiceRoom else { return }
        let room = roomId
        if !inVideoRoom {
            sfu?.destroy()
            sfu = nil
        }
        inVoiceRoom = false
        if let room, !inVideoRoom {
            try signaling.send(type: "leave_media", payload: ["roomId": room, "kind": "voice"])
            delegate?.rtcExpress(self, didLeaveVoiceRoom: room)
        }
    }

    public func joinVideoRoom() async throws {
        guard let roomId else { throw NSError(domain: "RTCExpress", code: 3) }
        guard let sfuUrl else { throw NSError(domain: "RTCExpress", code: 32) }
        if sfu == nil {
            sfu = SfuMediaEngine(sfuUrl: sfuUrl, userId: userId, authToken: token) { [weak self] type, payload in
                try self?.signaling.send(type: type, payload: payload)
            }
        }
        try await sfu?.joinRoom(roomId: roomId, options: SfuJoinOptions(audio: true, video: true, announceToRoom: true))
        inVoiceRoom = true
        inVideoRoom = true
        try signaling.send(type: "join_media", payload: ["roomId": roomId, "kind": "video"])
        delegate?.rtcExpress(self, didJoinVideoRoom: roomId)
    }

    public func leaveVideoRoom() throws {
        guard inVideoRoom else { return }
        let room = roomId
        sfu?.destroy()
        sfu = nil
        inVideoRoom = false
        inVoiceRoom = false
        if let room {
            try signaling.send(type: "leave_media", payload: ["roomId": room, "kind": "video"])
            delegate?.rtcExpress(self, didLeaveVideoRoom: room)
        }
    }

    public func callUser(_ peerUserId: String, video: Bool = false) throws {
        guard let roomId else { throw NSError(domain: "RTCExpress", code: 3) }
        if activeCall != nil { throw NSError(domain: "RTCExpress", code: 5, userInfo: [NSLocalizedDescriptionKey: "Already in a call"]) }
        let callId = UUID().uuidString
        let callType = video ? "video" : "voice"
        activeCall = ActiveCall(callId: callId, peerUserId: peerUserId, roomId: roomId, isCaller: true, callType: callType)
        if resolvedMediaMode() != "sfu" { ensureP2p(video: video) }
        try signaling.send(type: "call_invite", payload: [
            "roomId": roomId, "toUserId": peerUserId, "callId": callId, "callType": callType
        ])
        emitCallState("ringing")
    }

    public func acceptCall() throws {
        guard let call = activeCall else { throw NSError(domain: "RTCExpress", code: 4) }
        try signaling.send(type: "call_accept", payload: callPayload(call))
        if resolvedMediaMode() == "sfu" {
            Task { try? await startSfuCall(call) }
        } else {
            ensureP2p(video: call.callType == "video")
        }
        emitCallState("connecting")
    }

    public func rejectCall() throws {
        guard let call = activeCall else { return }
        try signaling.send(type: "call_reject", payload: callPayload(call))
        cleanupCall("rejected")
    }

    public func endCall() throws {
        guard let call = activeCall else { return }
        try signaling.send(type: "call_end", payload: callPayload(call))
        cleanupCall("ended")
    }

    public func startRecording() throws {
        if activeCall == nil && !inVoiceRoom && !inVideoRoom {
            throw NSError(domain: "RTCExpress", code: 22, userInfo: [NSLocalizedDescriptionKey: "Join a call or media room before recording"])
        }
        _ = try callRecorder.startRecording()
        delegate?.rtcExpressDidStartRecording(self)
    }

    public func stopRecording() throws -> RecordingResult {
        let result = try callRecorder.stopRecording()
        delegate?.rtcExpress(self, didStopRecording: result)
        return result
    }

    public var isRecording: Bool { callRecorder.isRecording }

    public func muteMicrophone(_ muted: Bool) { p2p?.muteMicrophone(muted) }
    public func muteCamera(_ muted: Bool) { p2p?.muteCamera(muted) }
    public func switchCamera() { p2p?.switchCamera() }

    public func destroy() {
        try? endCall()
        try? leaveVoiceRoom()
        try? leaveVideoRoom()
        signaling.close()
        p2p?.destroy()
        p2p = nil
    }

    func onConnected(userId: String) { delegate?.rtcExpress(self, didConnect: userId) }
    func onDisconnected() { delegate?.rtcExpressDidDisconnect(self) }
    func onRoomJoined(roomId: String, members: [String]) {
        delegate?.rtcExpress(self, didJoinRoom: roomId, members: members)
    }
    func onMessage(_ message: RoomMessage) { delegate?.rtcExpress(self, didReceive: message) }

    func onCallInvite(_ invite: CallInvite) {
        if activeCall != nil {
            try? signaling.send(type: "call_reject", payload: [
                "callId": invite.callId,
                "fromUserId": userId,
                "toUserId": invite.fromUserId,
                "roomId": invite.roomId,
                "callType": invite.callType
            ])
            return
        }
        activeCall = ActiveCall(callId: invite.callId, peerUserId: invite.fromUserId, roomId: invite.roomId, isCaller: false, callType: invite.callType)
        incomingCallNotifier?.onIncomingCall(invite, display: IncomingCallDisplay.from(invite))
        delegate?.rtcExpress(self, didReceive: invite)
        emitCallState("ringing")
    }

    func onCallAccept(_ payload: [String: Any]) {
        guard let call = activeCall else { return }
        if resolvedMediaMode() == "sfu" {
            Task {
                try? await startSfuCall(call)
                emitCallState("connected")
            }
            return
        }
        ensureP2p(video: call.callType == "video")
        p2p?.createOffer(peerUserId: call.peerUserId, callId: call.callId)
        emitCallState("connecting")
    }

    func onCallReject(_ payload: [String: Any]) { cleanupCall("rejected") }
    func onCallEnd(_ payload: [String: Any]) { cleanupCall("ended") }
    func onWebRtcOffer(_ payload: [String: Any]) {
        ensureP2p(video: activeCall?.callType == "video")
        p2p?.handleOffer(payload)
        emitCallState("connected")
    }
    func onWebRtcAnswer(_ payload: [String: Any]) {
        p2p?.handleAnswer(payload)
        emitCallState("connected")
    }
    func onIceCandidate(_ payload: [String: Any]) { p2p?.handleIceCandidate(payload) }
    func onSfuProducer(_ payload: [String: Any]) { sfu?.handleRemoteProducer(payload) }
    func onError(_ message: String, code: String?) {
        if code == "call_busy", activeCall?.isCaller == true { cleanupCall("rejected") }
        delegate?.rtcExpress(self, didError: message)
    }

    private func resolvedMediaMode() -> String {
        if mediaModePref == "p2p" { return "p2p" }
        if sfuUrl != nil { return "sfu" }
        return "p2p"
    }

    private func startSfuCall(_ call: ActiveCall) async throws {
        guard let sfuUrl else { throw NSError(domain: "RTCExpress", code: 32) }
        if sfu == nil {
            sfu = SfuMediaEngine(sfuUrl: sfuUrl, userId: userId, authToken: token) { [weak self] type, payload in
                try self?.signaling.send(type: type, payload: payload)
            }
        }
        let sfuRoomId = "\(call.roomId)-call-\(call.callId)"
        try await sfu?.joinRoom(
            roomId: sfuRoomId,
            options: SfuJoinOptions(audio: true, video: call.callType == "video", targetUserId: call.peerUserId, callId: call.callId)
        )
    }

    private func ensureP2p(video: Bool) {
        if p2p == nil {
            p2p = P2pMediaEngine(
                userId: userId,
                sendSignaling: { [weak self] type, payload in try self?.signaling.send(type: type, payload: payload) },
                activeCall: { [weak self] in
                    guard let call = self?.activeCall else { return nil }
                    return (call.callId, call.peerUserId)
                }
            )
            p2p?.onRemoteVideoTrack = { [weak self] track in
                guard let self else { return }
                self.delegate?.rtcExpress(self, remoteVideoTrack: track)
            }
        }
        p2p?.prepare(video: video)
    }

    private func callPayload(_ call: ActiveCall) -> [String: Any] {
        ["callId": call.callId, "fromUserId": userId, "toUserId": call.peerUserId, "roomId": call.roomId, "callType": call.callType]
    }

    private func emitCallState(_ state: String) {
        guard let call = activeCall else { return }
        delegate?.rtcExpress(self, callState: CallStateUpdate(
            callId: call.callId, state: state, peerUserId: call.peerUserId, roomId: call.roomId, callType: call.callType
        ))
    }

    private func cleanupCall(_ state: String) {
        emitCallState(state)
        if !inVoiceRoom && !inVideoRoom {
            sfu?.destroy()
            sfu = nil
        }
        p2p?.destroy()
        p2p = nil
        activeCall = nil
    }
}

File: packages/mobile-android/sdk/src/main/java/com/rtcexpress/sdk/RTCExpress.kt
package com.rtcexpress.sdk

import android.content.Context
import org.json.JSONObject
import org.webrtc.VideoTrack

class RTCExpress(private val context: Context) : SignalingClient.Listener {
    interface Listener {
        fun onConnected(userId: String) {}
        fun onDisconnected() {}
        fun onRoomJoined(roomId: String, members: List<String>) {}
        fun onMessage(message: RoomMessage) {}
        fun onCallInvite(invite: CallInvite) {}
        fun onCallState(update: CallStateUpdate) {}
        fun onRemoteVideo(track: VideoTrack) {}
        fun onVoiceRoomJoined(roomId: String) {}
        fun onVoiceRoomLeft(roomId: String) {}
        fun onVideoRoomJoined(roomId: String) {}
        fun onVideoRoomLeft(roomId: String) {}
        fun onRecordingStarted() {}
        fun onRecordingStopped(result: RecordingResult) {}
        fun onError(message: String) {}
    }

    private val signaling = SignalingClient()
    private val tokenClient = TokenClient()
    private val configClient = ConfigClient()
    private val historyClient = HistoryClient()
    private val callRecorder = CallRecorder(context)
    private var p2p: P2pMediaEngine? = null
    private var sfu: SfuMediaEngine? = null

    private var serverUrl = ""
    private var appId = ""
    private var token = ""
    private var userId = ""
    private var roomId: String? = null
    private var listener: Listener? = null
    private var incomingCallNotifier: IncomingCallNotifier? = null

    private var mediaModePref = "p2p"
    private var sfuUrl: String? = null
    private var inVoiceRoom = false
    private var inVideoRoom = false

    private var activeCall: ActiveCall? = null

    private data class ActiveCall(
        val callId: String,
        val peerUserId: String,
        val roomId: String,
        val isCaller: Boolean,
        val callType: String
    )

    fun setListener(listener: Listener?) {
        this.listener = listener
    }

    fun setIncomingCallNotifier(notifier: IncomingCallNotifier?) {
        this.incomingCallNotifier = notifier
    }

    fun init(options: RTCInitOptions) {
        serverUrl = options.serverUrl.trimEnd('/')
        appId = options.appId
        userId = options.userId
        token = options.token
        mediaModePref = options.mediaMode
        signaling.listener = this
        signaling.connect(serverUrl, options.token)
        runCatching {
            val config = configClient.fetchPlatformConfig(serverUrl, appId)
            sfuUrl = config.sfuUrl
        }
    }

    fun joinRoom(roomId: String) {
        this.roomId = roomId
        signaling.send("join_room", JSONObject().put("roomId", roomId))
    }

    fun sendMessage(text: String) {
        val room = roomId ?: throw IllegalStateException("Join a room first")
        signaling.send(
            "send_message",
            JSONObject().put("roomId", room).put("text", text)
        )
    }

    fun getMessageHistory(roomId: String, before: String? = null, limit: Int = 50): MessageHistoryPage {
        return historyClient.getMessageHistory(serverUrl, token, roomId, before, limit)
    }

    fun joinVoiceRoom() {
        val room = roomId ?: throw IllegalStateException("Join a room first")
        val url = sfuUrl ?: throw IllegalStateException("SFU not available")
        if (sfu == null) {
            sfu = SfuMediaEngine(
                context = context,
                sfuUrl = url,
                userId = userId,
                authToken = token,
                sendSignaling = { type, payload -> signaling.send(type, payload) },
                onRemoteVideo = { track -> listener?.onRemoteVideo(track) }
            )
        }
        sfu?.joinRoom(room, SfuJoinOptions(audio = true, video = false, announceToRoom = true))
        inVoiceRoom = true
        signaling.send("join_media", JSONObject().put("roomId", room).put("kind", "voice"))
        listener?.onVoiceRoomJoined(room)
    }

    fun leaveVoiceRoom() {
        if (!inVoiceRoom) return
        val room = roomId
        if (!inVideoRoom) {
            sfu?.destroy()
            sfu = null
        }
        inVoiceRoom = false
        if (room != null && !inVideoRoom) {
            signaling.send("leave_media", JSONObject().put("roomId", room).put("kind", "voice"))
            listener?.onVoiceRoomLeft(room)
        }
    }

    fun joinVideoRoom() {
        val room = roomId ?: throw IllegalStateException("Join a room first")
        val url = sfuUrl ?: throw IllegalStateException("SFU not available")
        if (sfu == null) {
            sfu = SfuMediaEngine(
                context = context,
                sfuUrl = url,
                userId = userId,
                authToken = token,
                sendSignaling = { type, payload -> signaling.send(type, payload) },
                onRemoteVideo = { track -> listener?.onRemoteVideo(track) }
            )
        }
        sfu?.joinRoom(room, SfuJoinOptions(audio = true, video = true, announceToRoom = true))
        inVoiceRoom = true
        inVideoRoom = true
        signaling.send("join_media", JSONObject().put("roomId", room).put("kind", "video"))
        listener?.onVideoRoomJoined(room)
    }

    fun leaveVideoRoom() {
        if (!inVideoRoom) return
        val room = roomId
        sfu?.destroy()
        sfu = null
        inVideoRoom = false
        inVoiceRoom = false
        if (room != null) {
            signaling.send("leave_media", JSONObject().put("roomId", room).put("kind", "video"))
            listener?.onVideoRoomLeft(room)
        }
    }

    fun callUser(peerUserId: String, video: Boolean = false) {
        val room = roomId ?: throw IllegalStateException("Join a room first")
        if (activeCall != null) throw IllegalStateException("Already in a call")
        val callId = JsonUtil.newCallId()
        val callType = if (video) "video" else "voice"
        activeCall = ActiveCall(callId, peerUserId, room, true, callType)
        if (resolvedMediaMode() == "sfu") {
            signaling.send(
                "call_invite",
                JSONObject()
                    .put("roomId", room)
                    .put("toUserId", peerUserId)
                    .put("callId", callId)
                    .put("callType", callType)
            )
            emitCallState("ringing", peerUserId, room, callId, callType)
            return
        }
        ensureP2p(video)
        signaling.send(
            "call_invite",
            JSONObject()
                .put("roomId", room)
                .put("toUserId", peerUserId)
                .put("callId", callId)
                .put("callType", callType)
        )
        emitCallState("ringing", peerUserId, room, callId, callType)
    }

    fun acceptCall() {
        val call = activeCall ?: throw IllegalStateException("No incoming call")
        signaling.send("call_accept", callPayload(call, call.peerUserId))
        if (resolvedMediaMode() == "sfu") {
            startSfuCall(call)
        } else {
            ensureP2p(call.callType == "video")
        }
        emitCallState("connecting", call.peerUserId, call.roomId, call.callId, call.callType)
    }

    fun rejectCall() {
        val call = activeCall ?: return
        signaling.send("call_reject", callPayload(call, call.peerUserId))
        cleanupCall("rejected")
    }

    fun endCall() {
        val call = activeCall ?: return
        signaling.send("call_end", callPayload(call, call.peerUserId))
        cleanupCall("ended")
    }

    fun startRecording() {
        if (activeCall == null && !inVoiceRoom && !inVideoRoom) {
            throw IllegalStateException("Join a call or media room before recording")
        }
        callRecorder.startRecording()
        listener?.onRecordingStarted()
    }

    fun stopRecording(): RecordingResult {
        val result = callRecorder.stopRecording()
        listener?.onRecordingStopped(result)
        return result
    }

    fun isRecording() = callRecorder.isRecording

    fun muteMicrophone(muted: Boolean) {
        sfu?.muteMicrophone(muted) ?: p2p?.muteMicrophone(muted)
    }

    fun muteCamera(muted: Boolean) {
        sfu?.muteCamera(muted) ?: p2p?.muteCamera(muted)
    }

    fun switchCamera() {
        sfu?.switchCamera() ?: p2p?.switchCamera()
    }

    fun destroy() {
        endCall()
        leaveVoiceRoom()
        leaveVideoRoom()
        signaling.close()
        p2p?.destroy()
        p2p = null
    }

    companion object {
        fun fetchToken(serverUrl: String, request: TokenRequest): TokenResponse {
            return TokenClient().fetchToken(serverUrl, request)
        }
    }

    override fun onConnected(userId: String) {
        listener?.onConnected(userId)
    }

    override fun onDisconnected() {
        listener?.onDisconnected()
    }

    override fun onRoomJoined(roomId: String, members: List<String>) {
        listener?.onRoomJoined(roomId, members)
    }

    override fun onMessage(message: RoomMessage) {
        listener?.onMessage(message)
    }

    override fun onCallInvite(invite: CallInvite) {
        if (activeCall != null) {
            signaling.send(
                "call_reject",
                JSONObject()
                    .put("callId", invite.callId)
                    .put("fromUserId", userId)
                    .put("toUserId", invite.fromUserId)
                    .put("roomId", invite.roomId)
                    .put("callType", invite.callType)
            )
            return
        }
        activeCall = ActiveCall(
            invite.callId,
            invite.fromUserId,
            invite.roomId,
            false,
            invite.callType
        )
        incomingCallNotifier?.onIncomingCall(invite, IncomingCallDisplay.from(invite))
        listener?.onCallInvite(invite)
        emitCallState("ringing", invite.fromUserId, invite.roomId, invite.callId, invite.callType)
    }

    override fun onCallAccept(payload: JSONObject) {
        val call = activeCall ?: return
        if (resolvedMediaMode() == "sfu") {
            startSfuCall(call)
            emitCallState("connected", call.peerUserId, call.roomId, call.callId, call.callType)
            return
        }
        ensureP2p(call.callType == "video")
        p2p?.createOffer(call.peerUserId, call.callId)
        emitCallState("connecting", call.peerUserId, call.roomId, call.callId, call.callType)
    }

    override fun onCallReject(payload: JSONObject) {
        cleanupCall("rejected")
    }

    override fun onCallEnd(payload: JSONObject) {
        cleanupCall("ended")
    }

    override fun onWebRtcOffer(payload: JSONObject) {
        ensureP2p(activeCall?.callType == "video")
        p2p?.handleOffer(payload)
        activeCall?.let {
            emitCallState("connected", it.peerUserId, it.roomId, it.callId, it.callType)
        }
    }

    override fun onWebRtcAnswer(payload: JSONObject) {
        p2p?.handleAnswer(payload)
        activeCall?.let {
            emitCallState("connected", it.peerUserId, it.roomId, it.callId, it.callType)
        }
    }

    override fun onWebRtcIceCandidate(payload: JSONObject) {
        p2p?.handleIceCandidate(payload)
    }

    override fun onSfuProducer(payload: JSONObject) {
        sfu?.handleRemoteProducer(payload)
    }

    override fun onError(message: String, code: String?) {
        if (code == "call_busy" && activeCall?.isCaller == true) {
            cleanupCall("rejected")
        }
        listener?.onError(message)
    }

    private fun resolvedMediaMode(): String {
        if (mediaModePref == "p2p") return "p2p"
        if (mediaModePref == "sfu" && !sfuUrl.isNullOrBlank()) return "sfu"
        if (mediaModePref == "auto" && !sfuUrl.isNullOrBlank()) return "sfu"
        return "p2p"
    }

    private fun startSfuCall(call: ActiveCall) {
        val url = sfuUrl ?: throw IllegalStateException("SFU not available")
        if (sfu == null) {
            sfu = SfuMediaEngine(
                context = context,
                sfuUrl = url,
                userId = userId,
                authToken = token,
                sendSignaling = { type, payload -> signaling.send(type, payload) },
                onRemoteVideo = { track -> listener?.onRemoteVideo(track) }
            )
        }
        val sfuRoomId = "${call.roomId}-call-${call.callId}"
        sfu?.joinRoom(
            sfuRoomId,
            SfuJoinOptions(
                audio = true,
                video = call.callType == "video",
                targetUserId = call.peerUserId,
                callId = call.callId
            )
        )
    }

    private fun ensureP2p(video: Boolean) {
        if (p2p == null) {
            p2p = P2pMediaEngine(
                context = context,
                userId = userId,
                sendSignaling = { type, payload -> signaling.send(type, payload) },
                activeCall = {
                    activeCall?.let { it.callId to it.peerUserId }
                }
            ).also {
                it.onRemoteVideoTrack = { track -> listener?.onRemoteVideo(track) }
            }
        }
        p2p?.prepare(video)
    }

    private fun callPayload(call: ActiveCall, toUserId: String): JSONObject {
        return JSONObject()
            .put("callId", call.callId)
            .put("fromUserId", userId)
            .put("toUserId", toUserId)
            .put("roomId", call.roomId)
            .put("callType", call.callType)
    }

    private fun emitCallState(state: String, peerUserId: String, roomId: String, callId: String, callType: String) {
        listener?.onCallState(CallStateUpdate(callId, state, peerUserId, roomId, callType))
    }

    private fun cleanupCall(state: String) {
        val call = activeCall ?: return
        emitCallState(state, call.peerUserId, call.roomId, call.callId, call.callType)
        if (!inVoiceRoom && !inVideoRoom) {
            sfu?.destroy()
            sfu = null
        }
        p2p?.destroy()
        p2p = null
        activeCall = null
    }
}
